cmd/clef, signer: security fixes

holiman opened the pull request:
Replaces https://github.com/ethereum/go-ethereum/pull/17427 . 

* Remove local path info when communicating externally (still available in internal channel) 
* Adds `Origin` and `User-Agent` to the context, and makes it visible for Clef
* Audit fix: document that external fields are set by remote caller
* Audit fix: harden storage kv-store, so values cannot be swapped around
* Audit fix: By default reject on tx validation errors, have `--advanced` mode to enable warnings instead of rejections
* Audit fix: Implement password complexity check on new accounts
* Audit fix: remove import functionality from external interface

karalabe commented on signer/core/cliui.go:
```diff
@@ -113,6 +113,9 @@ func (ui *CommandlineUI) ApproveTx(request *SignTxRequest) (SignTxResponse, erro
 	fmt.Printf("value: %v wei\n", weival)
+	fmt.Printf("gas: %v\n", request.Transaction.Gas)
+	fmt.Printf("gasprice: %v wei\n", request.Transaction.GasPrice.ToInt())
+	fmt.Printf("nonce: %v\n", request.Transaction.Nonce)
```
Prev fields were aligned. I assume you want to keep them so.

holiman replied:
fixed

holiman commented:
I've now modified it a bit. The cli ui, when performing a listing, shows this (full info):
```
------- Signer info -------
* extapi_http : n/a
* extapi_ipc : /home/user/.clef/clef.ipc
* extapi_version : 3.0.0
* intapi_version : 2.0.0
-------- List Account request--------------
A request has been made to list all accounts. 
You can select which accounts the caller can see
  [x] 0x8A8eAFb1cf62BfBeb1741769DAE1a9dd47996192
    URL: keystore:///home/user/.ethereum/keystore/UTC--2018-05-18T18-24-52.475027848Z--8a8eafb1cf62bfbeb1741769dae1a9dd47996192
    Type: Account
  [x] 0x64f951c2C1B289638b5c7FB6eb6657ee359Dd984
    URL: keystore:///home/user/.ethereum/keystore/UTC--2018-08-17T08-48-34.811873860Z--64f951c2c1b289638b5c7fb6eb6657ee359dd984
    Type: Account

```
I also added `User-Agent` and `Origin` to the `ctx` provided by  `http.Server`. 
Making a request:
```
[user@work clef]$ curl -H "Content-Type: application/json" -X POST --data-binary @signtx.json http://localhost:8550 -i
```
This is how it looks:
```
--------- Transaction request-------------
to:    0x07a565b7ed7d7a678680a4c162885bedbb695fe0

WARNING: Invalid checksum on to-address!

from:     0xcf6cd422f9b45778ad8a564edfbf8abab96e363a [chksum INVALID]
value:    16 wei
gas:      0x333 (819)
gasprice: 291 wei
nonce:    0x0 (0)
data:     0x4401a6e40000000000000000000000000000000000000000000000000000000000000012

Transaction validation:
  * WARNING : Invalid checksum on to-address
  * Info : safeSend(address: 0x0000000000000000000000000000000000000012)


Request context:
	127.0.0.1:44728 -> HTTP/1.1 -> localhost:8550

	User-Agent: curl/7.53.1
	Origin: 
-------------------------------------------
Approve? [y/N]:
```

fjl commented:
This looks good, just a small request: Please improve the error message for invalid characters in password. It shouldn't list all possible characters.

ligi commented on signer/storage/aes_gcm_storage.go:
```diff
@@ -129,7 +129,10 @@ func (s *AESEncryptedStorage) writeEncryptedStorage(creds map[string]storedCrede
 }
 
-func encrypt(key []byte, plaintext []byte) ([]byte, []byte, error) {
+// encrypt encrypts plaingtext with the given key, with additionaldata
```
There is one g too much in plaintext and one space missing between additional and data

holiman replied:
fixed

holiman commented:
Fixed!

fjl commented:
Sorry, I merged a typo fix PR that made this one unmergable.

holiman commented:
rebased
